examples_test/config_test.go

```diff
@@ -0,0 +1,144 @@
+package examples_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/lyraproj/hiera/hiera"
+	"github.com/lyraproj/hiera/hieraapi"
+	"github.com/lyraproj/hiera/provider"
+	"github.com/lyraproj/pcore/px"
+	"github.com/lyraproj/pcore/types"
+)
+
+// sayHello is a very simple "lookup_key" function that just returns the result of concatenating
+// the key with the string " world".
+func sayHello(pc hieraapi.ServerContext, key string) px.Value {
+	return types.WrapString(key + ` world`)
+}
+
+// TestConfig_hardwired utilizes Hiera in the simplest way possible. No configuration file and no options. Just
+// a function performs a lookup of a key.
+func TestConfig_hardwired(t *testing.T) {
+	// Use the hiera.DoWithParent to initialize a Hiera context (session) with the sayHello as the top-level function and
+	// perform a lookup.
+	//
+	// The DoWithParent is meant to be called once and the created context can then be used for any number of lookups that
+	// uses the same configuration. The sessions life-cycle can be compared to the compilers life-cycle in puppet.
+	hiera.DoWithParent(context.Background(), sayHello, nil, func(c px.Context) {
+		result := hiera.Lookup(hiera.NewInvocation(c, nil, nil), `hello`, nil, nil)
+		if result == nil || `hello world` != result.String() {
+			t.Fatalf("unexpected result %v", result)
+		}
+	})
+}
+
+// TestHelloWorld_semiHardWired uses the special "lookup_key" function MuxLookupKey. This function use the configuration
+// option LookupProvidersKey where it expects to find a slice of "lookup_key" functions to use.
+func TestConfig_semiHardWired(t *testing.T) {
+	// Create "global" options for this Hiera session (global as in "kept in the top-level context". An executable may
+	// have an unlimited set of such contexts and there are no global variables involved.
```

Comment:
This is hard to grasp - it mentions "global", "session" "context" and "executable". How are "global" options for a session different from 'non "global"' ?
Is this just saying that you can have as many independent sessions as you like?